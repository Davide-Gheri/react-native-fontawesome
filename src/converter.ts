import React from 'react';
import { AbstractElement, SizeProp } from '@fortawesome/fontawesome-svg-core';
import {
  Svg,
  Path,
  Defs,
  Rect,
  Mask,
  ClipPath,
  G,
  Circle
} from 'react-native-svg';

export const mapping: { [key: string]: React.ComponentClass<any, any> } = {
  svg: Svg,
  path: Path,
  rect: Rect,
  g: G,
  defs: Defs,
  mask: Mask,
  clipPath: ClipPath,
  circle: Circle
};

export function sizeToInt(size?: SizeProp): number {
  const base = 24;
  switch (true) {
    case size === 'xs':
      return base * 0.75;
    case size === 'lg':
      return (base * 4) / 3;
    case size === 'sm':
      return base * 0.875;
    case typeof size === 'string' &&
      (size as string).indexOf('x') === (size as string).length - 1:
      const multi = parseInt((size as string).replace('x', ''), 10);
      return base * multi;
    default:
      return base;
  }
}

export function convert(
  tree: AbstractElement,
  otherProps: { size?: SizeProp; color?: string } = {},
  key?: string | number
) {
  const childrens: any = tree.children
    ? tree.children.map((t, k) =>
        convert(t, { color: otherProps.color }, k + 1)
      )
    : [];

  const { size, color } = otherProps;

  const props = Object.assign(
    {},
    tree.attributes,
    key ? { key } : null,
    color ? { fill: color } : null,
    {
      width: sizeToInt(size).toString(),
      height: sizeToInt(size).toString()
    }
  );

  return React.createElement(mapping[tree.tag], props, childrens);
}
