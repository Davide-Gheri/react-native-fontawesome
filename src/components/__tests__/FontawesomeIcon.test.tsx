import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { library, IconName, SizeProp, icon as iconRn } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';
import Icon, { IconProps } from '../FontawesomeIcon';
import log from '../../logger';
import { sizeToInt } from '../../converter';

jest.mock('../../logger', () => jest.fn());

library.add(fas);

const render = (props: IconProps): ReactTestRendererJSON | null => renderer.create(<Icon {...props}/>).toJSON();

describe('<FontawesomeIcon>', () => {

  it('should render correctly passing an icon name', () => {
    const icon = render({icon: 'circle'});

    expect(icon).toMatchSnapshot();
  });

  it('should render correctly passing an icon array', () => {
    const icon = render({icon: ['far', 'circle']});

    expect(icon).toMatchSnapshot();
  });

  it('should render correctly passing an icon object', () => {
    const icon = render({icon: {prefix: 'far', iconName: 'circle'}});

    expect(icon).toMatchSnapshot();
  });

  it('should render a masked icon passing a mask icon', () => {
    const icon = render({icon: 'abacus', mask: 'circle'});

    expect(icon).toMatchSnapshot();
  });

  it('should render null if icon is not found', () => {
    const icon = render({icon: 'notAnIcon' as IconName});

    expect(icon).toMatchInlineSnapshot('null');
    expect(log).toHaveBeenCalledWith('Could not find icon', { prefix: 'fas', iconName: 'notAnIcon' });
  });

  it('should render null if no icon is passed', () => {
    const icon = render({icon: null} as any);

    expect(icon).toMatchInlineSnapshot('null');
    expect(log).toHaveBeenCalledWith('Could not find icon', null);
  });

  it('should add styles to the wrapper View component', () => {
    const style = {
      marginLeft: 20,
      marginRight: 20,
    };
    const icon = render({style, icon: 'circle'}) as ReactTestRendererJSON;

    expect(icon.props).toEqual({style});
  });

  it('should set SVG width and height based on size prop', () => {

    (['xs', 'lg', 'sm', '1x', '2x', '3x', '4x', '5x', '6x', '7x', '8x', '9x', '10x'] as SizeProp[])
    .forEach((size) => {
      const toInt = sizeToInt(size);
      const icon = render({size, icon: 'circle'})  as ReactTestRendererJSON;
      const SVG = (icon.children as ReactTestRendererJSON[])[0];

      expect(SVG.props.width).toBe(toInt);
      expect(SVG.props.height).toBe(toInt);
    });
  });

  it('should set SVG fill based on color prop', () => {
    const icon = render({icon: 'circle', color: '#fafafa'}) as ReactTestRendererJSON;
    const SVG = (icon.children as ReactTestRendererJSON[])[0];

    expect(SVG.props.fill).toEqual('#fafafa');
  });

  it('should apply transform from fontawesome transform keys', () => {
    const transform = {
      flipX: false,
      flipY: false,
      rotate: 15,
      size: 56,
      x: -4,
      y: 0,
    };

    const icn = iconRn({iconName: 'circle', prefix: 'fas'}, {transform});
    const style = icn.abstract[0].attributes.style;

    const icon = render({transform, icon: 'circle'}) as ReactTestRendererJSON;
    const SVG = (icon.children as ReactTestRendererJSON[])[0];

    expect(SVG.props.style.some((s: any) => s === style)).toBeTruthy();
  });
});
