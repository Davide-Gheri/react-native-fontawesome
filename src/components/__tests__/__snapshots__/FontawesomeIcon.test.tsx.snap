// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FontawesomeIcon> should render a masked icon passing a mask icon 1`] = `null`;

exports[`<FontawesomeIcon> should render correctly passing an icon array 1`] = `null`;

exports[`<FontawesomeIcon> should render correctly passing an icon name 1`] = `
<View>
  <RNSVGSvgView
    align="xMidYMid"
    aria-hidden="true"
    bbHeight="24"
    bbWidth="24"
    class="svg-inline--fa fa-circle fa-w-16"
    data-icon="circle"
    data-prefix="fas"
    focusable="false"
    height="24"
    meetOrSlice={0}
    minX={0}
    minY={0}
    role="img"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        undefined,
        null,
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={512}
    vbWidth={512}
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={
        Array [
          0,
          0,
          0,
          0,
          1,
        ]
      }
      fillOpacity={1}
      fillRule={1}
      font={Object {}}
      matrix={
        Array [
          1,
          0,
          -0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "fillOpacity",
          "fillRule",
          "stroke",
          "strokeWidth",
          "strokeOpacity",
          "strokeDasharray",
          "strokeDashoffset",
          "strokeLinecap",
          "strokeLinejoin",
          "strokeMiterlimit",
        ]
      }
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={null}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={1}
      x={0}
      y={0}
    >
      <RNSVGPath
        d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
        fill={
          Array [
            2,
          ]
        }
        fillOpacity={1}
        fillRule={1}
        matrix={
          Array [
            1,
            0,
            -0,
            1,
            0,
            0,
          ]
        }
        opacity={1}
        originX={0}
        originY={0}
        propList={
          Array [
            "fill",
          ]
        }
        rotation={0}
        scaleX={1}
        scaleY={1}
        skewX={0}
        skewY={0}
        stroke={null}
        strokeDasharray={null}
        strokeDashoffset={null}
        strokeLinecap={0}
        strokeLinejoin={0}
        strokeMiterlimit={4}
        strokeOpacity={1}
        strokeWidth={1}
        x={0}
        y={0}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`<FontawesomeIcon> should render correctly passing an icon object 1`] = `null`;
